process/sync/baseForkDetector.go

```diff
@@ -118,7 +118,8 @@ func (bfd *baseForkDetector) removeInvalidReceivedHeaders() {
 		for i := 0; i < len(hdrInfos); i++ {
 			roundDif := int64(hdrInfos[i].round) - int64(finalCheckpointRound)
 			nonceDif := int64(hdrInfos[i].nonce) - int64(finalCheckpointNonce)
-			isReceivedHeaderInvalid := hdrInfos[i].state == process.BHReceived && roundDif < nonceDif
+			hasStateReceived := hdrInfos[i].state == process.BHReceived || hdrInfos[i].state == process.BHReceivedTooLate
+			isReceivedHeaderInvalid := hasStateReceived && roundDif < nonceDif
```

Comment:
it is clearly malicious behavious if there are more nonces than rounds.

Reply:
If you received an invalid header from a malicious node, it can be. Also when you are syncing and you are receiving headers with a higher nonce, you wouldn't know at that time if they are valid or not in terms of nonce/round differences. But while you are syncing your finalCheckPoint would advance and you can detect now that some of received headers with higher nonces would be invalid and could be deleted form this fork list. This is just a protection against flooding if these headers would escape somehow from interceptors checks. 

One simple example is that let say you are syncing and the chronology is in round 10. You just committed header with round 2 and none 2. If you would received now header with round 10 and nonce 3 you can not say that this header is invalid. But if you would sync the next header with round 4 and nonce 3 and then header with round 6 and nonce 4 which will make final the header with round 4 and nonce 3, now its clear that the header with round 10 and nonce 3 is invalid and you can delete it.

Reply:
OK!